internal/refactoring/move_execute.go

```diff
@@ -39,13 +43,18 @@ func ApplyMoves(stmts []MoveStatement, state *states.State) MoveResults {
 
 	g := buildMoveStatementGraph(stmts)
 
-	// If there are any cycles in the graph then we'll not take any action
-	// at all. The separate validation step should detect this and return
-	// an error.
-	if len(g.Cycles()) != 0 {
+	// If the graph is not valid the we will not take any action at all. The
+	// separate validation step should detect this and return an error.
+	if diags := validateMoveStatementGraph(g); diags.HasErrors() {
+		log.Printf("[ERROR] ApplyMoves: %s", diags.ErrWithWarnings())
 		return ret
 	}
 
+	// The graph must be reduced in order for ReverseDepthFirstWalk to work
+	// correctly, since it is built from following edges and can skip over
+	// dependencies if there is a direct edge to a transitive dependency.
+	g.TransitiveReduction()
+
```

Comment:
The need to do this was surprising to me, because I just assumed (my first mistake!) that `ReverseDepthFirstWalk` would be doing something like Kahn's algorithm where transitive reduction would only be a marginal performance improvement, rather than a crucial part of its correctness.

Now that we know this, it might be helpful to at least document that prerequisite in [the `AcyclicGraph.ReverseDepthFirstWalk`](https://pkg.go.dev/github.com/hashicorp/terraform/internal/dag#AcyclicGraph.ReverseDepthFirstWalk) documentation, so there's at least a little smaller chance of a future maintainer making the same mistake!

(That doc comment also has the typical confusion from across the `dag` package about what exactly "up" and "down" mean, but that's a broader problem for another day...)

Reply:
Talking of documentation: I notice that [`AcyclicGraph.TransitiveReduction`](https://pkg.go.dev/github.com/hashicorp/terraform/internal/dag#AcyclicGraph.TransitiveReduction)'s comment asserts that it only guarantees correct behavior for a graph that would pass [`AcyclicGraph.Validate`](https://pkg.go.dev/github.com/hashicorp/terraform/internal/dag#AcyclicGraph.Validate), which includes the requirement of having only a single "root" node.

You noted that we're explicitly not using `Validate` in order to avoid the need for a single root. Looking at the implementation of `TransitiveReduction` it seems to use [`AcyclicGraph.DepthFirstWalk`](https://pkg.go.dev/github.com/hashicorp/terraform/internal/dag#AcyclicGraph.DepthFirstWalk) internally, which seems to use a similar algorithm as `AcyclicGraph.DepthFirstWalk` and so gives me some concern about why it's acceptable to use a depth-first walk _as part of_ computing the transitive reduction if we're also asserting that transitive reduction is required for correctness of these walks.

It seems like the documentation commentary in `dag` is all a bit suspect, to be honest, so I'm not meaning to suggest that there's anything definitively _wrong_ with what you did here, but the fact that the existing commentary and implementation seems to disagree with the assumptions this new code is making is giving me some pause; perhaps it would be worth figuring out if the `dag` package comments are actually correct, and then updating them if not, or changing our approach in this new caller if they _are_ correct. :thinking:

Reply:
I didn't want to go into auditing the `dag` package, since it has a number of peculiarities based upon how it has always been used in terraform. That assertion in `TransitiveReduction` is overly conservative, but it seemed safe, especially since our primary use is the concurrent walk which does require a single root. I took our use of a walk from multiple roots as the exception to how the dag is normally handled (this is the first time we've ever not enforced a complete graph have a single root) and verified the behavior based on what the refactoring package required. The walk functions really only accept multiple roots for use in the Ancestors/Descendents collection, which is unordered. 

Our naive DFS walk functions are largely unchanged since the package was made, and they don't do a full topological sort -- they walk the edges presented immediately and rely on a previous `TransitiveReduction` to ensure an overall sorted order. This has always been fine because we don't ever process a graph in-order without a reducing it first.

I'm going to follow up with a small PR for the `dag` package to clarify things, since I'd like to remove a couple functions left from the last refactoring too, and don't want to muddy up the changes here.